File: analyzer/parser.py
"""R20 workbook parser + validator for PRETAG AMIS.

Handles both shapes seen in real files:
  * a flat Regional R20 (one sheet of members)
  * a zone workbook (one sheet per district)

Absorbs the quirks catalogued in docs/build-plan.html section 2:
dirty sheet names, ghost rows, blank rows, trailing empty columns, unused
Sheet2/Sheet3, numeric employee-number cells, double spaces in names.
"""
from __future__ import annotations
import hashlib, os, re
from dataclasses import dataclass, field

import openpyxl

from .mapping import Mapping, norm_text

REQUIRED_HEADERS = ["EMPLOYEE NO", "NAME OF EMPLOYEE", "MANAGEMENT UNIT", "DISTRICT", "REGION"]
REGION_CANON = "Ashanti"


def _clean_name(s: str) -> str:
    return re.sub(r"\s+", " ", norm_text(s)).strip()


def _emp_to_text(v) -> str:
    if v is None:
        return ""
    if isinstance(v, float) and v.is_integer():
        v = int(v)
    return re.sub(r"\s+", "", str(v)).strip()


@dataclass
class Row:
    sheet: str
    excel_row: int
    employee_no: str
    name: str
    management_unit: str
    raw_district: str
    raw_region: str
    district: str | None = None
    zone: str | None = None
    problems: list[str] = field(default_factory=list)


@dataclass
class ParseResult:
    path: str
    file_hash: str
    sheets_read: list[str]
    sheets_skipped: list[str]
    rows: list[Row]
    blank_rows_dropped: int

    # ---- derived views -------------------------------------------------
    @property
    def valid(self) -> list[Row]:
        return [r for r in self.rows if not r.problems]

    @property
    def duplicates(self) -> dict[str, list[Row]]:
        by_emp: dict[str, list[Row]] = {}
        for r in self.rows:
            if r.employee_no:
                by_emp.setdefault(r.employee_no, []).append(r)
        return {e: rs for e, rs in by_emp.items() if len(rs) > 1}

    @property
    def missing_employee_no(self) -> list[Row]:
        return [r for r in self.rows if not r.employee_no]

    @property
    def unmapped(self) -> dict[str, int]:
        out: dict[str, int] = {}
        for r in self.rows:
            if r.zone is None and r.raw_district:
                out[r.raw_district] = out.get(r.raw_district, 0) + 1
        return dict(sorted(out.items(), key=lambda kv: -kv[1]))

    def summary(self) -> dict:
        return {
            "file": os.path.basename(self.path),
            "file_hash": self.file_hash[:12],
            "sheets_read": len(self.sheets_read),
            "sheets_skipped": self.sheets_skipped,
            "total_member_rows": len(self.rows),
            "valid_rows": len(self.valid),
            "blank_rows_dropped": self.blank_rows_dropped,
            "missing_employee_no": len(self.missing_employee_no),
            "duplicate_employee_nos": len(self.duplicates),
            "unmapped_districts": len(self.unmapped),
            "unmapped_detail": self.unmapped,
            "status": self._status(),
        }

    def _status(self) -> str:
        if self.missing_employee_no or self.unmapped or self.duplicates:
            return "NEEDS REVIEW"
        return "VALIDATED"


def _header_index(cells) -> bool:
    got = [norm_text(c).upper() for c in list(cells)[:5]]
    return got == REQUIRED_HEADERS


def parse(path: str, mapping: Mapping | None = None) -> ParseResult:
    mapping = mapping or Mapping()
    with open(path, "rb") as fh:
        file_hash = hashlib.sha256(fh.read()).hexdigest()

    wb = openpyxl.load_workbook(path, read_only=True, data_only=True)
    rows: list[Row] = []
    read, skipped = [], []
    blank_dropped = 0

    for ws in wb.worksheets:
        data = list(ws.iter_rows(values_only=True))
        if not data or not _header_index(data[0]):
            if norm_text(ws.title):
                skipped.append(ws.title)
            continue
        read.append(ws.title)
        for i, raw in enumerate(data[1:], start=2):
            cells = list(raw)[:5] + [None] * (5 - len(list(raw)[:5]))
            if all(c is None or str(c).strip() == "" for c in cells):
                blank_dropped += 1
                continue
            row = Row(
                sheet=norm_text(ws.title),
                excel_row=i,
                employee_no=_emp_to_text(cells[0]),
                name=_clean_name(cells[1]),
                management_unit=norm_text(cells[2]),
                raw_district=norm_text(cells[3]),
                raw_region=norm_text(cells[4]),
            )
            if not row.employee_no:
                row.problems.append("missing employee number")
            if not row.name:
                row.problems.append("missing name")
            d = mapping.resolve(row.raw_district)
            if d:
                row.district, row.zone = d.name, d.zone
            else:
                row.problems.append(f"unmapped district: {row.raw_district!r}")
            if row.raw_region and row.raw_region.lower() != REGION_CANON.lower():
                row.problems.append(f"unexpected region: {row.raw_region!r}")
            rows.append(row)

    wb.close()

    # duplicate employee numbers within the file
    seen: dict[str, Row] = {}
    for r in rows:
        if not r.employee_no:
            continue
        if r.employee_no in seen:
            r.problems.append("duplicate employee number")
            first = seen[r.employee_no]
            if "duplicate employee number" not in first.problems:
                first.problems.append("duplicate employee number")
        else:
            seen[r.employee_no] = r

    return ParseResult(path, file_hash, read, skipped, rows, blank_dropped)

File: lib/r20/parser.ts
/**
 * R20 workbook parser + validator. TypeScript port of analyzer/parser.py.
 * Handles both a flat Regional R20 and a multi-sheet zone workbook, and absorbs
 * the quirks in docs/build-plan.html section 2.
 */
import ExcelJS from "exceljs";
import { cleanName, employeeNoToText, normText } from "./normalize";
import type { Resolver } from "./mapping";

export const REQUIRED_HEADERS = [
  "EMPLOYEE NO",
  "NAME OF EMPLOYEE",
  "MANAGEMENT UNIT",
  "DISTRICT",
  "REGION",
];
const REGION_CANON = "ashanti";

export type ParsedRow = {
  sheet: string;
  excelRow: number;
  employeeNo: string;
  name: string;
  managementUnit: string;
  rawDistrict: string;
  rawRegion: string;
  districtId: number | null;
  zoneId: number | null;
  status: "valid" | "warning" | "error";
  message: string | null;
};

export type ParseResult = {
  sheetsRead: string[];
  sheetsSkipped: string[];
  blankRowsDropped: number;
  rows: ParsedRow[];
};

export type ValidationSummary = {
  totalRows: number;
  validRows: number;
  blankRowsDropped: number;
  missingEmployeeNo: number;
  duplicateEmployeeNos: number;
  unmappedRows: number;
  unmappedDistricts: { district: string; count: number }[];
  sheetsRead: string[];
  sheetsSkipped: string[];
  status: "validated" | "needs_review";
};

function cellText(cell: ExcelJS.Cell | undefined): unknown {
  if (!cell) return null;
  const v = cell.value;
  if (v && typeof v === "object") {
    if ("text" in v) return (v as { text: string }).text;
    if ("result" in v) return (v as { result: unknown }).result;
    if ("richText" in v)
      return (v as { richText: { text: string }[] }).richText.map((r) => r.text).join("");
  }
  return v;
}

// How many rows / leading columns to probe when locating the header. Some R20s
// carry a leading serial-number column or a blank/title row above the headers.
const HEADER_SCAN_ROWS = 15;
const HEADER_SCAN_COLS = 4;

function matchHeaderAt(row: ExcelJS.Row, colOffset: number): boolean {
  for (let i = 0; i < 5; i++) {
    if (normText(cellText(row.getCell(colOffset + 1 + i))).toUpperCase() !== REQUIRED_HEADERS[i]) {
      return false;
    }
  }
  return true;
}

/** Locate the R20 header: returns {headerRow, colOffset} or null. */
function findHeader(ws: ExcelJS.Worksheet): { headerRow: number; colOffset: number } | null {
  const last = Math.min(HEADER_SCAN_ROWS, ws.rowCount || HEADER_SCAN_ROWS);
  for (let r = 1; r <= last; r++) {
    const row = ws.getRow(r);
    for (let c = 0; c <= HEADER_SCAN_COLS; c++) {
      if (matchHeaderAt(row, c)) return { headerRow: r, colOffset: c };
    }
  }
  return null;
}

export async function parseWorkbook(buffer: ArrayBuffer | Buffer): Promise<ParseResult> {
  const wb = new ExcelJS.Workbook();
  // ExcelJS accepts a Node Buffer or ArrayBuffer; the cast sidesteps a
  // @types/node Buffer-generic mismatch.
  await wb.xlsx.load(buffer as unknown as ArrayBuffer);

  const rows: ParsedRow[] = [];
  const sheetsRead: string[] = [];
  const sheetsSkipped: string[] = [];
  let blankRowsDropped = 0;

  wb.eachSheet((ws) => {
    const title = normText(ws.name);
    const header = findHeader(ws);
    if (!header) {
      if (title) sheetsSkipped.push(ws.name);
      return;
    }
    sheetsRead.push(title);
    const { headerRow, colOffset } = header;
    const cols = [1, 2, 3, 4, 5].map((n) => colOffset + n);

    ws.eachRow((row, rowNumber) => {
      if (rowNumber <= headerRow) return;
      const values = cols.map((c) => cellText(row.getCell(c)));
      if (values.every((v) => v === null || v === undefined || String(v).trim() === "")) {
        blankRowsDropped += 1;
        return;
      }
      rows.push({
        sheet: title,
        excelRow: rowNumber,
        employeeNo: employeeNoToText(values[0]),
        name: cleanName(values[1]),
        managementUnit: normText(values[2]),
        rawDistrict: normText(values[3]),
        rawRegion: normText(values[4]),
        districtId: null,
        zoneId: null,
        status: "valid",
        message: null,
      });
    });
  });

  return { sheetsRead, sheetsSkipped, blankRowsDropped, rows };
}

/** Attach mapping + validation to parsed rows in place, and return the summary. */
export function validate(parsed: ParseResult, resolver: Resolver): ValidationSummary {
  const problems = (r: ParsedRow, msg: string, level: "warning" | "error" = "error") => {
    r.status = r.status === "error" ? "error" : level;
    r.message = r.message ? `${r.message}; ${msg}` : msg;
  };

  const seen = new Map<string, ParsedRow>();
  const unmapped = new Map<string, number>();

  for (const r of parsed.rows) {
    if (!r.employeeNo) problems(r, "missing employee number");
    if (!r.name) problems(r, "missing name", "warning");

    const d = r.rawDistrict ? resolver.resolve(r.rawDistrict) : null;
    if (d) {
      r.districtId = d.districtId;
      r.zoneId = d.zoneId;
    } else {
      problems(r, `unmapped district: ${r.rawDistrict || "(blank)"}`);
      unmapped.set(r.rawDistrict, (unmapped.get(r.rawDistrict) ?? 0) + 1);
    }

    if (r.rawRegion && r.rawRegion.toLowerCase() !== REGION_CANON) {
      problems(r, `unexpected region: ${r.rawRegion}`, "warning");
    }

    if (r.employeeNo) {
      const prev = seen.get(r.employeeNo);
      if (prev) {
        problems(r, "duplicate employee number");
        problems(prev, "duplicate employee number");
      } else {
        seen.set(r.employeeNo, r);
      }
    }
  }

  const missingEmployeeNo = parsed.rows.filter((r) => !r.employeeNo).length;
  const dupCount = new Set(
    parsed.rows.filter((r) => r.message?.includes("duplicate employee number")).map((r) => r.employeeNo),
  ).size;
  const unmappedRows = [...unmapped.values()].reduce((a, b) => a + b, 0);
  const validRows = parsed.rows.filter((r) => r.status === "valid").length;

  return {
    totalRows: parsed.rows.length,
    validRows,
    blankRowsDropped: parsed.blankRowsDropped,
    missingEmployeeNo,
    duplicateEmployeeNos: dupCount,
    unmappedRows,
    unmappedDistricts: [...unmapped.entries()]
      .map(([district, count]) => ({ district, count }))
      .sort((a, b) => b.count - a.count),
    sheetsRead: parsed.sheetsRead,
    sheetsSkipped: parsed.sheetsSkipped,
    status:
      missingEmployeeNo || unmappedRows || dupCount ? "needs_review" : "validated",
  };
}
